ESQL: Replace AssertionError in mappers

alex-spies opened the pull request:
We want to avoid throwing errors; AssertionErrors are thrown in mappers whenever an operation is not supported for whatever reason. Throw UnsupportedOperationException instead.

alex-spies committed:
Replace AssertionError in mappers

We want to avoid throwing errors; AssertionErrors are thrown in mappers
whenever an operation is not supported for whatever reason. Throw
UnsupportedOperationException instead.

elasticsearchmachine commented:
Pinging @elastic/es-ql (Team:QL)

elasticsearchmachine commented:
Pinging @elastic/elasticsearch-esql (:Query Languages/ES|QL)

alex-spies commented:
@elasticsearchmachine run elasticsearch-ci/bwc

bpintea commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/convert/AbstractConvertFunction.java:
```diff
@@ -43,7 +43,7 @@ protected Supplier<EvalOperator.ExpressionEvaluator> evaluator(Supplier<EvalOper
         var evaluator = evaluators().get(sourceType);
         if (evaluator == null) {
-            throw new AssertionError("unsupported type [" + sourceType + "]");
+            throw new UnsupportedOperationException("unsupported type [" + sourceType + "]");
```
We have some QL Exceptions, that split in sever and client and `EsqlIllegalArgumentException` might be a good choice for these ones.

alex-spies replied:
This looks more appropriate to me, as it seems to be intended for server-side errors that should've been handled during validation.

@luigidellaquila suggested `UnsupportedOperationException` [here](https://github.com/elastic/elasticsearch/pull/98352#pullrequestreview-1580262191), but if we're all happy with `EsqlIllegalArgumentException` I'm going to go by your suggestion instead.

bpintea reviewed (approved):
LGTM

costin reviewed (approved):
Use EsqlIllegalArgumentException - it's a specific error to ESQL so it can be properly caught and is REST friendly.
UOE and IAE are generic exception, they are thrown even by the JDK and thus cannot be easily be traced.
The compute module uses them to avoid depending on QL however inside *QL we should throw our own exceptions.

alex-spies commented:
> Use EsqlIllegalArgumentException - it's a specific error to ESQL so it can be properly caught and is REST friendly. UOE and IAE are generic exception, they are thrown even by the JDK and thus cannot be easily be traced. The compute module uses them to avoid depending on QL however inside *QL we should throw our own exceptions.

Alright, I'll use `EsqlIllegalArgumentException` then! For this to fully make sense, I think I'll add another commit on top that replaces some `UnsupportedOperationException`s so we consistently use `EsqlIllegalArgumentException`.

alex-spies committed:
Replace by EsqlIllegalArgumentException instead

This can be handled better in REST request in case we e.g. encounter
expression mappings with unsupported data types during query planning
(which normally should be already caught during query verification).

alex-spies committed:
Remove UnsupportedOperationException in esql

Replace by EsqlIllegalArgumentException which behaves better during REST
requests.

alex-spies commented:
Alright, now this PR introduces consistent usage of `EsqlIllegalArgumentException` across the the `esql` package, except for compute and tests.

Since this touches a lot more places now, a re-review would be good @costin or @bpintea . If you think this makes sense, I'll refactor using one or two new helper functions because a lot of time the `EsqlIllegalArgumentException`s (formerly: `UnsupportedOperationException`s) are thrown in a very similar pattern.
